app/src/app/store/mapRenderSubs.ts

```diff
@@ -128,15 +136,29 @@ export const getRenderSubscriptions = (useMapStore: typeof _useMapStore) => {
             .filter(([key, value]) => value !== null)
             .map(([key]) => key)
         );
-        setLockedFeatures(new Set(nonNullZones));
+        setLockedFeatures(nonNullZones);
         // now unlocked, was previously locked
       } else if (Array.isArray(lockPaintedAreas)) {
+        const previousWasArray = Array.isArray(prevLockPaintedAreas);
         const nonNullZones = new Set(
           [...zoneAssignments.entries()]
-            .filter(([key, value]) => lockPaintedAreas.includes(value))
+            .filter(
+              ([key, value]) =>
+                // locked zones include assignment zone
+                lockPaintedAreas.includes(value) ||
+                // locked zones are the same, and this individual feature was previously locked
+                (sameLockedAreas && lockedFeatures.has(key)) ||
+                // locked zones are changed, BUT this individual feature is not in a zone
+                // that was previously locked
+                (!sameLockedAreas &&
+                  previousWasArray &&
+                  !lockPaintedAreas.includes(value) &&
+                  !prevLockPaintedAreas.includes(value) &&
+                  lockedFeatures.has(key))
```

Comment:
PP: Are all of these sets? I wonder if you could use boolean logic–i.e. `intersection`, `difference`, `union`–instead of destructuring `zoneAssignments` and iterating through each element. Anyway probably negligible perf benefits if any

Reply:
PS. Remembering that `zoneAssignments` is a Map not a Set (duh!) so you can ignore the part of my comment about destructuring `zoneAssignments` as that would have to be done either way

Reply:
(And since this is an active bug we should merge and can always come back to this if it's hurting perf, which I doubt it is)